File: packages/environments/src/tunnel.ts
import { randomUUID } from "node:crypto";
import { appendFileSync, existsSync, readFileSync } from "node:fs";
import type { Server } from "node:http";
import { join } from "node:path";
import { PassThrough } from "node:stream";
import {
  type Context,
  context,
  decode,
  fail,
  HubError,
  hash,
} from "@contremaitre/execution/context";
import { atomicWrite } from "@contremaitre/execution/files";
import { EnvironmentLocks } from "@contremaitre/execution/locks";
import { run } from "@contremaitre/execution/process";
import { closeServer, type Lookup, listen, proxyServer } from "@contremaitre/routing/proxy";
import { Schema } from "effect";
import type { Manager } from "./manager.js";
import type { Environment, TunnelReservation } from "./model.js";
import { providerConfig } from "./tunnel-provider.js";

const readinessSchema = Schema.Struct({ version: Schema.Literal(2), ready: Schema.Boolean });
const responseSchema = Schema.Struct({
  version: Schema.Literal(1),
  reservation_id: Schema.optional(Schema.String),
  url: Schema.optional(Schema.String),
  ready: Schema.optional(Schema.Boolean),
  capabilities: Schema.optional(Schema.Record({ key: Schema.String, value: Schema.Boolean })),
});
interface Connector {
  controller: AbortController;
  done: Promise<void>;
  server: Server;
  input: PassThrough;
  ready: boolean;
}
export class Tunnels {
  private readonly failures = new Map<string, HubError>();
  failure(id: string, name: string) {
    return this.failures.get(`${id}/${name}`);
  }
  reset(env: Environment) {
    for (const name of Object.keys(env.Services))
      this.failures.delete(`${env.Identity.ID}/${name}`);
  }
  private readonly active = new Map<string, Connector>();
  private readonly locks = new EnvironmentLocks();
  private closing = false;
  private readonly machine: string;
  constructor(
    readonly manager: Manager,
    readonly lookup: Lookup,
  ) {
    const path = join(manager.store.home, "tunnel-machine-id");
    if (!existsSync(path)) atomicWrite(path, randomUUID());
    this.machine = readFileSync(path, "utf8").trim();
    if (!/^[a-f0-9-]{36}$/.test(this.machine)) fail("Invalid tunnel machine identity");
  }
  private identity(env: Environment) {
    return {
      environment_id: hash(`${this.machine}\0${env.Identity.ID}`).slice(0, 32),
      machine_id: this.machine,
      project: env.Identity.Project,
      branch: env.Identity.Branch,
      workspace_id: hash(env.Identity.Workspace).slice(0, 32),
    };
  }
  running(id: string, name: string) {
    return this.active.get(`${id}/${name}`)?.ready ?? false;
  }
  private config(provider?: string) {
    return (
      providerConfig(this.manager.store.home, provider) ??
      fail("Run contremaitre tunnel to set up the SaaS provider first")
    );
  }
  private async call(
    ctx: Context,
    env: Environment,
    name: string,
    operation: string,
    provider?: string,
  ) {
    const cfg = this.config(provider),
      reservation = env.tunnels?.[name];
    try {
      return decode(
        responseSchema,
        JSON.parse(
          (
            await run(ctx, [cfg.executable, operation], {
              stdin: Buffer.from(
                JSON.stringify({
                  version: 1,
                  config: cfg.config,
                  ...this.identity(env),
                  service_id: name,
                  display_name: env.Identity.Name,
                  reservation_id: reservation?.ID,
                }),
              ),
              timeout: 30_000,
              maxOutput: 1048576,
              strictOutput: true,
              stderr: () => {},
            })
          ).toString(),
        ),
        "tunnel provider response",
      );
    } catch (error) {
      const permanent =
        error instanceof HubError &&
        ([77, 78].includes(error.exitCode ?? 0) ||
          (!error.exitCode && error.classification === "permanent"));
      fail(`Tunnel provider ${operation} failed`, permanent ? "permanent" : "transient");
    }
  }
  async reserve(
    ctx: Context,
    env: Environment,
    name: string,
    provider?: string,
  ): Promise<TunnelReservation> {
    return this.locks.use([`${env.Identity.ID}/${name}`], ctx.signal, async () => {
      if (this.closing) fail("Hub is shutting down");
      if (!env.Services[name]?.HTTP) fail(`Service ${name} is not HTTP`);
      env.tunnels ??= {};
      let reservation = env.tunnels[name];
      if (reservation && provider && reservation.Provider !== provider)
        fail("Tunnel reservation belongs to another provider; release it before switching");
      const cfg = this.config(reservation?.Provider ?? provider);
      const caps = await this.call(ctx, env, name, "capabilities", cfg.name);
      if (
        !caps.capabilities?.stable_urls ||
        !caps.capabilities.https ||
        !caps.capabilities.foreground_sessions
      )
        fail(
          "Provider must support stable_urls, https and foreground_sessions; upgrade the tunnel adapter",
        );
      if (!reservation) {
        const reserved = await this.call(ctx, env, name, "reserve", cfg.name);
        let url: URL;
        try {
          url = new URL(reserved.url ?? "");
        } catch {
          return fail("Provider returned invalid reservation");
        }
        if (url.protocol !== "https:" || url.username || url.password || !reserved.reservation_id)
          fail("Provider returned invalid reservation");
        reservation = {
          Provider: cfg.name,
          ID: reserved.reservation_id,
          URL: url.toString(),
          Desired: false,
          Connected: false,
        };
        env.tunnels[name] = reservation;
        this.manager.save();
      }
      return reservation;
    });
  }
  async start(
    ctx: Context,
    env: Environment,
    name: string,
    lease: { id: string; expires: number; enabled: () => boolean },
  ): Promise<TunnelReservation> {
    const reservation = await this.reserve(ctx, env, name);
    return this.locks.use([`${env.Identity.ID}/${name}`], ctx.signal, async () => {
      if (this.closing) fail("Hub is shutting down");
      const key = `${env.Identity.ID}/${name}`;
      if (this.active.has(key)) fail("Connector already running");
      const cfg = this.config(reservation.Provider);
      const publicHost = new URL(reservation.URL).host;
      const localHost = new URL(this.manager.localURL(env, name)).host;
      const server = proxyServer(this.lookup, () => {
        const route = this.lookup(key);
        return route
          ? { ...route, upstream: lease.enabled() ? route.upstream : "", publicHost, localHost }
          : undefined;
      });
      const port = await listen(server, 0);
      const controller = new AbortController();
      const input = new PassThrough();
      const log = join(this.manager.store.home, `tunnel-${env.Identity.ID}-${name}.log`);
      atomicWrite(log, "");
      let logged = 0,
        line = Buffer.alloc(0),
        reported = false;
      let readyResolve: () => void = () => {},
        readyReject: (error: Error) => void = () => {};
      const ready = new Promise<void>((resolve, reject) => {
        readyResolve = resolve;
        readyReject = reject;
      });
      let exited = false;
      const done = run(
        {
          ...context(controller.signal),
          processDirectory: join(this.manager.store.home, "processes"),
        },
        [cfg.executable, "start"],
        {
          stdin: input,
          /* initial request is written below, followed by lease renewals */
          timeout: 2_147_483_647,
          stdout: (chunk) => {
            if (reported) return;
            const end = chunk.indexOf(10);
            const segment = end < 0 ? chunk : chunk.subarray(0, end);
            if (line.length + segment.length > 65536) {
              reported = true;
              readyReject(Error("Provider readiness response exceeded limit"));
              return;
            }
            line = Buffer.concat([line, segment]);
            if (end >= 0) {
              reported = true;
              try {
                const response = decode(
                  readinessSchema,
                  JSON.parse(line.toString()),
                  "provider readiness",
                );
                if (!response.ready) fail("Provider did not report ready");
                readyResolve();
              } catch {
                readyReject(
                  new HubError({
                    message: "Provider did not report ready",
                    classification: "permanent",
                  }),
                );
              }
            }
          },
          stderr: (chunk) => {
            const data = chunk.subarray(0, Math.max(0, 10 * 1048576 - logged));
            if (data.length) {
              appendFileSync(log, data);
              logged += data.length;
            }
          },
        },
      )
        .then(
          () => readyReject(Error("Tunnel connector exited")),
          (error) => {
            const permanent = error instanceof HubError && [77, 78].includes(error.exitCode ?? 0);
            const failure = new HubError({
              message: permanent
                ? "Tunnel provider refused authorization or configuration"
                : "Tunnel connector failed",
              classification: permanent ? "permanent" : "transient",
            });
            if (permanent) this.failures.set(key, failure);
            readyReject(failure);
          },
        )
        .finally(() => {
          exited = true;
          input.destroy();
          this.active.delete(key);
          return closeServer(server);
        });
      const connector: Connector = { controller, done, server, input, ready: false };
      this.active.set(key, connector);
      input.write(
        `${JSON.stringify({
          version: 2,
          config: cfg.config,
          ...this.identity(env),
          service_id: name,
          display_name: env.Identity.Name,
          reservation_id: reservation.ID,
          upstream: `http://127.0.0.1:${port}`,
          session_id: lease.id,
          service_ids: Object.keys(env.Services)
            .filter((n) => env.Services[n].HTTP)
            .sort(),
          expires_at: lease.expires,
          enabled: false,
        })}\n`,
      );
      const timeout = setTimeout(() => readyReject(Error("Provider readiness timed out")), 30_000);
      const abort = () => readyReject(Error("Tunnel start cancelled"));
      ctx.signal.addEventListener("abort", abort, { once: true });
      if (ctx.signal.aborted) abort();
      try {
        await ready;
        if (exited) fail("Tunnel connector exited before readiness completed");
        controller.signal.throwIfAborted();
        connector.ready = true;
        reservation.Desired = true;
        reservation.Connected = true;
        this.manager.save();
        return { ...reservation };
      } catch (e) {
        controller.abort();
        await done;
        throw e;
      } finally {
        clearTimeout(timeout);
        ctx.signal.removeEventListener("abort", abort);
      }
    });
  }
  async stop(ctx: Context, env: Environment, name: string, release: boolean) {
    await this.locks.use([`${env.Identity.ID}/${name}`], ctx.signal, async () => {
      const reservation = env.tunnels?.[name];
      if (!reservation) return;
      reservation.Desired = false;
      reservation.Connected = false;
      this.manager.save();
      const active = this.active.get(`${env.Identity.ID}/${name}`);
      if (active) {
        await closeServer(active.server);
        active.controller.abort();
        await active.done;
      }
      await this.call(ctx, env, name, release ? "release" : "stop", reservation.Provider);
      if (release) delete env.tunnels?.[name];
      this.manager.save();
    });
  }
  renew(env: Environment, id: string, expires: number, enabled: boolean) {
    for (const name of Object.keys(env.tunnels ?? {})) {
      const connector = this.active.get(`${env.Identity.ID}/${name}`);
      if (!connector) continue;
      if (
        !connector.input.write(
          `${JSON.stringify({ version: 2, operation: "renew", session_id: id, expires_at: expires, enabled })}\n`,
        )
      )
        connector.controller.abort();
    }
  }
  async shutdown() {
    this.closing = true;
    for (const c of this.active.values()) c.controller.abort();
    await Promise.all([...this.active.values()].map((c) => c.done));
  }
}

File: internal/hub/tunnel.go
package hub

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/url"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"sync"
	"syscall"
	"time"
)

// Providers are installed local programs. JSON goes through stdin/stdout, never a shell.
type ProviderConfig struct {
	Executable string          `json:"executable"`
	Config     json.RawMessage `json:"config,omitempty"`
}
type TunnelReservation struct {
	Provider  string
	ID        string
	URL       string
	Desired   bool
	Connected bool
}
type ProviderRequest struct {
	Version       int             `json:"version"`
	Config        json.RawMessage `json:"config,omitempty"`
	EnvironmentID string          `json:"environment_id,omitempty"`
	ServiceID     string          `json:"service_id,omitempty"`
	DisplayName   string          `json:"display_name,omitempty"`
	ReservationID string          `json:"reservation_id,omitempty"`
	Upstream      string          `json:"upstream,omitempty"`
}
type ProviderResponse struct {
	Version      int             `json:"version"`
	ID           string          `json:"reservation_id"`
	URL          string          `json:"url"`
	Ready        bool            `json:"ready"`
	Capabilities map[string]bool `json:"capabilities"`
}
type providerProcess struct {
	cmd  *exec.Cmd
	done chan struct{}
	log  *os.File
}
type TunnelManager struct {
	mu     sync.Mutex
	Home   string
	active map[string]*providerProcess
}

func NewTunnelManager(home string) *TunnelManager {
	return &TunnelManager{Home: home, active: map[string]*providerProcess{}}
}
func (t *TunnelManager) config(provider string) (ProviderConfig, error) {
	var all struct {
		Default   string                    `json:"default"`
		Providers map[string]ProviderConfig `json:"providers"`
	}
	b, e := os.ReadFile(filepath.Join(t.Home, "tunnels.json"))
	if e != nil {
		return ProviderConfig{}, fmt.Errorf("configure a tunnel provider in %s: %w", filepath.Join(t.Home, "tunnels.json"), e)
	}
	if e = json.Unmarshal(b, &all); e != nil {
		return ProviderConfig{}, e
	}
	if provider == "" {
		provider = all.Default
	}
	p, ok := all.Providers[provider]
	if !ok || p.Executable == "" {
		return p, fmt.Errorf("tunnel provider %q is not configured", provider)
	}
	if !filepath.IsAbs(p.Executable) {
		return p, fmt.Errorf("provider executable must be an absolute path")
	}
	return p, nil
}
func (t *TunnelManager) defaultName() (string, error) {
	var cfg struct {
		Default string `json:"default"`
	}
	b, e := os.ReadFile(filepath.Join(t.Home, "tunnels.json"))
	if e != nil {
		return "", e
	}
	e = json.Unmarshal(b, &cfg)
	return cfg.Default, e
}
func providerCall(ctx context.Context, cfg ProviderConfig, operation string, req ProviderRequest) (ProviderResponse, error) {
	req.Version = 1
	req.Config = cfg.Config
	b, e := json.Marshal(req)
	if e != nil {
		return ProviderResponse{}, e
	}
	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
	defer cancel()
	cmd := exec.CommandContext(ctx, cfg.Executable, operation)
	cmd.Stdin = bytes.NewReader(b)
	stderr := cappedBuffer{Limit: 64 * 1024}
	cmd.Stderr = &stderr
	stdout, e := cmd.StdoutPipe()
	if e != nil {
		return ProviderResponse{}, e
	}
	if e = cmd.Start(); e != nil {
		return ProviderResponse{}, e
	}
	out, e := io.ReadAll(io.LimitReader(stdout, 1024*1024+1))
	if e != nil || len(out) > 1024*1024 {
		_ = cmd.Process.Kill()
		_ = cmd.Wait()
		return ProviderResponse{}, fmt.Errorf("provider response exceeded limit or failed")
	}
	if e = cmd.Wait(); e != nil {
		return ProviderResponse{}, fmt.Errorf("provider %s failed: %w", operation, e)
	}
	var result ProviderResponse
	e = json.Unmarshal(out, &result)
	if e == nil && result.Version != 1 {
		e = fmt.Errorf("unsupported provider protocol")
	}
	return result, e
}
func (t *TunnelManager) Start(ctx context.Context, env *Environment, service, upstream string, save func() error) (*TunnelReservation, error) {
	t.mu.Lock()
	defer t.mu.Unlock()
	key := env.Identity.ID + "/" + service
	if p := t.active[key]; p != nil {
		select {
		case <-p.done:
			delete(t.active, key)
		default:
			return env.Tunnels[service], nil
		}
	}
	if !validUpstream(upstream) {
		return nil, fmt.Errorf("tunnel upstream must be loopback HTTP")
	}
	r := env.Tunnels[service]
	provider := ""
	var e error
	if r != nil {
		provider = r.Provider
	} else {
		provider, e = t.defaultName()
		if e != nil {
			return nil, e
		}
	}
	cfg, e := t.config(provider)
	if e != nil {
		return nil, e
	}
	req := ProviderRequest{Version: 1, EnvironmentID: env.Identity.ID, ServiceID: service, DisplayName: env.Identity.Name}
	caps, e := providerCall(ctx, cfg, "capabilities", req)
	if e != nil {
		return nil, e
	}
	if !caps.Capabilities["stable_urls"] || !caps.Capabilities["https"] {
		return nil, fmt.Errorf("provider must support stable_urls and https")
	}
	if r == nil {
		res, e := providerCall(ctx, cfg, "reserve", req)
		if e != nil {
			return nil, e
		}
		u, e := url.Parse(res.URL)
		if e != nil || u.Scheme != "https" || u.Host == "" || u.User != nil || res.ID == "" {
			return nil, fmt.Errorf("provider returned invalid reservation")
		}
		r = &TunnelReservation{Provider: provider, ID: res.ID, URL: res.URL}
		if env.Tunnels == nil {
			env.Tunnels = map[string]*TunnelReservation{}
		}
		env.Tunnels[service] = r
		if e = save(); e != nil {
			return nil, e
		}
	}
	req.ReservationID = r.ID
	req.Upstream = upstream
	req.Config = cfg.Config
	b, _ := json.Marshal(req)
	cmd := exec.Command(cfg.Executable, "start")
	cmd.Stdin = bytes.NewReader(b)
	pipe, e := cmd.StdoutPipe()
	if e != nil {
		return nil, e
	}
	logPath := filepath.Join(t.Home, "tunnel-"+env.Identity.ID+"-"+service+".log")
	log, e := os.OpenFile(logPath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0600)
	if e != nil {
		return nil, e
	}
	cmd.Stderr = &limitedLog{Writer: log, Remaining: 10 * 1024 * 1024}
	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
	if e = cmd.Start(); e != nil {
		log.Close()
		return nil, e
	}
	ready := make(chan error, 1)
	go func() {
		reader := bufio.NewReaderSize(pipe, 64*1024)
		line, e := reader.ReadSlice('\n')
		if e == nil {
			var event ProviderResponse
			e = json.Unmarshal(line, &event)
			if e == nil && (!event.Ready || event.Version != 1) {
				e = fmt.Errorf("provider did not report ready")
			}
		}
		ready <- e
		_, _ = io.Copy(io.Discard, reader)
	}()
	select {
	case e = <-ready:
	case <-ctx.Done():
		e = ctx.Err()
	case <-time.After(30 * time.Second):
		e = fmt.Errorf("provider readiness timed out")
	}
	if e != nil {
		_ = syscall.Kill(-cmd.Process.Pid, syscall.SIGKILL)
		_ = cmd.Wait()
		log.Close()
		return nil, e
	}
	p := &providerProcess{cmd: cmd, done: make(chan struct{}), log: log}
	t.active[key] = p
	go func() { _ = cmd.Wait(); log.Close(); close(p.done) }()
	r.Desired = true
	return r, save()
}
func (t *TunnelManager) Stop(ctx context.Context, env *Environment, service string, release bool) error {
	t.mu.Lock()
	defer t.mu.Unlock()
	r := env.Tunnels[service]
	if r == nil {
		return nil
	}
	r.Desired = false
	key := env.Identity.ID + "/" + service
	if p := t.active[key]; p != nil {
		_ = syscall.Kill(-p.cmd.Process.Pid, syscall.SIGTERM)
		select {
		case <-p.done:
		case <-time.After(5 * time.Second):
			_ = syscall.Kill(-p.cmd.Process.Pid, syscall.SIGKILL)
			<-p.done
		}
		delete(t.active, key)
	}
	cfg, e := t.config(r.Provider)
	if e != nil {
		return e
	}
	operation := "stop"
	if release {
		operation = "release"
	}
	_, e = providerCall(ctx, cfg, operation, ProviderRequest{ReservationID: r.ID, EnvironmentID: env.Identity.ID, ServiceID: service})
	if e != nil {
		return e
	}
	r.Desired = false
	if release {
		delete(env.Tunnels, service)
	}
	return nil
}
func (t *TunnelManager) Running(envID, service string) bool {
	t.mu.Lock()
	defer t.mu.Unlock()
	p := t.active[envID+"/"+service]
	if p == nil {
		return false
	}
	select {
	case <-p.done:
		return false
	default:
		return true
	}
}
func (t *TunnelManager) Close() {
	t.mu.Lock()
	defer t.mu.Unlock()
	for key, p := range t.active {
		_ = syscall.Kill(-p.cmd.Process.Pid, syscall.SIGTERM)
		select {
		case <-p.done:
		case <-time.After(5 * time.Second):
			_ = syscall.Kill(-p.cmd.Process.Pid, syscall.SIGKILL)
		}
		delete(t.active, key)
	}
}
func validUpstream(s string) bool {
	u, e := url.Parse(s)
	return e == nil && u.Scheme == "http" && strings.HasPrefix(u.Host, "127.0.0.1:")
}

type limitedLog struct {
	Writer    io.Writer
	Remaining int
}

func (l *limitedLog) Write(p []byte) (int, error) {
	n := len(p)
	if l.Remaining <= 0 {
		return n, nil
	}
	if len(p) > l.Remaining {
		p = p[:l.Remaining]
	}
	written, e := l.Writer.Write(p)
	l.Remaining -= written
	if e != nil {
		return written, e
	}
	return n, nil
}
